File: wallet-kit/src/main/java/bitcoin/walllet/kit/struct/OutPoint.java
package bitcoin.walllet.kit.struct;

import com.fasterxml.jackson.databind.annotation.JsonSerialize;

import java.io.IOException;

import bitcoin.walllet.kit.common.io.BitcoinInput;
import bitcoin.walllet.kit.common.io.BitcoinOutput;
import bitcoin.walllet.kit.common.serializer.HashSerializer;

public class OutPoint {

    /**
     * 32-bytes, the hash of the referenced transaction.
     */
    @JsonSerialize(using = HashSerializer.class)
    public byte[] hash;

    /**
     * uint32, the index of the specific output in the transaction. The first
     * output is 0, etc.
     */
    public long index;

    public OutPoint(BitcoinInput input) throws IOException {
        this.hash = input.readBytes(32);
        this.index = input.readUnsignedInt();
    }

    public byte[] toByteArray() {
        return new BitcoinOutput().write(hash).writeUnsignedInt(index).toByteArray();
    }
}

File: wallet-kit/src/main/kotlin/bitcoin/wallet/kit/models/OutPoint.kt
package bitcoin.wallet.kit.models

import bitcoin.walllet.kit.io.BitcoinInput
import bitcoin.walllet.kit.io.BitcoinOutput
import io.realm.RealmObject
import java.io.IOException

/**
 * An outpoint describes the transaction output that is connected to a transaction input.
 * It consists of the hash for the transaction containing the output
 * and the index of the output within the transaction.
 */
open class OutPoint : RealmObject {

    // 32-bytes, the hash of the referenced transaction.
    // @JsonSerialize(using = HashSerializer::class)
    var hash: ByteArray = byteArrayOf()

    // Uint32, the index of the specific output in the transaction. The first output is 0, etc.
    var index: Long = 0

    constructor()

    @Throws(IOException::class) constructor(input: BitcoinInput) {
        this.hash = input.readBytes(32)
        this.index = input.readUnsignedInt()
    }

    fun toByteArray(): ByteArray {
        return BitcoinOutput()
                .write(hash)
                .writeUnsignedInt(index)
                .toByteArray()
    }
}
